File: Graphs/Cycledetectiondirected.java
import java.util.ArrayList;

public class Cycledetectiondirected {

    static class Edge {
        int src;
        int dest;

        public Edge(int s, int d) {
            this.src = s;
            this.dest = d;
        }
    }

    static void createGraph(ArrayList<Edge> graph[]){
        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<Edge>();
        }
        graph[0].add(new Edge(0, 1));
        graph[2].add(new Edge(2, 1));
        graph[2].add(new Edge(2, 3));
        graph[3].add(new Edge(3, 4));
        graph[4].add(new Edge(4, 2));
    }

    public static void findNeighbour(ArrayList<Edge> graph[], int index){
        for (int i = 0; i < graph[index].size(); i++){
            Edge e = graph[index].get(i);
            System.out.println("Neighbour of " + e.src + " is " + e.dest);
        }
    }

    public static boolean isCycleDetected(ArrayList<Edge> graph[],boolean vis[],int curr,boolean rec[]){
        vis[curr] = true; rec[curr] = true;
        for (int i = 0; i < graph[curr].size(); i++) {
            Edge e = graph[curr].get(i);
            if(rec[e.dest]){
                return true;
            }
            else if(!vis[e.dest]){
                if(isCycleDetected(graph, vis, e.dest, rec)){
                    return true;
                }
            }
        }
            
        rec[curr] = false;
        return false;
    }

    public static void main(String[] args) {
        /*
         0 ---> 1 <--- 2 <---
                      /      \
                     -> 3 --> 4

        0       1     2       3       4
        0,1          2,1     3,4      4,2
                     2,3      
         */

        int V = 5;
        ArrayList<Edge> graph [] =  new ArrayList[V];
        createGraph(graph);
        boolean ans = isCycleDetected(graph, new boolean[V], V, new boolean[V]);
        System.out.println(ans);
        System.out.println("done");

        //  for (int j = 0; j < V; j++) {   
        //      findNeighbour(graph, j);
        //  }
    }
}

File: Graphs/Cycledetectiondirected.py
class Edge:
    def __init__(self, src, dest):
        self.src = src
        self.dest = dest

def create_graph(graph):
    for i in range(len(graph)):
        graph[i] = []

    graph[0].append(Edge(0, 1))
    graph[2].append(Edge(2, 1))
    graph[2].append(Edge(2, 3))
    graph[3].append(Edge(3, 4))
    graph[4].append(Edge(4, 2))

def print_all_neighbours(graph, index):
    for i in graph[index]:
        print(i.dest)

def isCycleUtil(graph, curr, vis, stack):
    vis[curr] = True
    stack[curr] = True
    for i in range(len(graph[curr])):
        e = graph[curr][i] 
        if stack[e.dest]:
            return True
        elif not vis[e.dest] and isCycleUtil(graph, e.dest,vis, stack):
                return True
    stack[curr] = False
    return False


def isCyclic(graph):
    vis = [False for _ in range(len(graph))]
    for i in range(len(graph)):
        if not vis[i]:
            cycle = isCycleUtil(graph, i , vis, [False for _ in range(len(vis))])
            if(cycle):
                return True
    return False

def main():

    #   0 ---> 1 <--- 2 <---
    #                 /      \
    #                -> 3 --> 4
    #
    #      0       1     2       3       4
    #      0,1          2,1     3,4      4,2
    #                   2,3  

    V = 5
    graph = [[] for _ in range(V)]
    create_graph(graph)
    print(isCyclic(graph))

    """
    for j in range(V):
        print(f"Neighbours of {j} is") 
        print_all_neighbours(graph, j)
        print()
     """

if __name__ == "__main__":
    main()